File: src/main/java/org/radarbase/appserver/service/FcmNotificationService.java
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */

package org.radarbase.appserver.service;

import java.time.Instant;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

import org.radarbase.appserver.converter.NotificationConverter;
import org.radarbase.appserver.dto.fcm.FcmNotificationDto;
import org.radarbase.appserver.dto.fcm.FcmNotifications;
import org.radarbase.appserver.entity.Notification;
import org.radarbase.appserver.entity.Project;
import org.radarbase.appserver.entity.User;
import org.radarbase.appserver.event.state.MessageState;
import org.radarbase.appserver.event.state.dto.NotificationStateEventDto;
import org.radarbase.appserver.exception.InvalidNotificationDetailsException;
import org.radarbase.appserver.exception.InvalidUserDetailsException;
import org.radarbase.appserver.exception.NotFoundException;
import org.radarbase.appserver.exception.NotificationAlreadyExistsException;
import org.radarbase.appserver.repository.NotificationRepository;
import org.radarbase.appserver.repository.ProjectRepository;
import org.radarbase.appserver.repository.UserRepository;
import org.radarbase.appserver.service.scheduler.NotificationSchedulerService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * {@link Service} for interacting with the {@link Notification} {@link javax.persistence.Entity}
 * using the {@link NotificationRepository}.
 *
 * @author yatharthranjan
 */
@SuppressWarnings("PMD.DataflowAnomalyAnalysis")
@Service
public class FcmNotificationService implements NotificationService {

    // TODO Add option to specify a scheduling provider (default will be fcm)
    // TODO: Use strategy pattern for handling notifications for scheduling and adding to database

    private static final String INVALID_SUBJECT_ID_MESSAGE =
            "The supplied Subject ID is invalid. No user found. Please Create a User First.";
    private final transient NotificationRepository notificationRepository;
    private final transient UserRepository userRepository;
    private final transient ProjectRepository projectRepository;
    private final transient NotificationSchedulerService schedulerService;
    private final transient NotificationConverter notificationConverter;
    private final transient ApplicationEventPublisher notificationStateEventPublisher;

    @Autowired
    public FcmNotificationService(
            NotificationRepository notificationRepository,
            UserRepository userRepository,
            ProjectRepository projectRepository,
            NotificationSchedulerService schedulerService,
            NotificationConverter notificationConverter,
            ApplicationEventPublisher eventPublisher) {
        this.notificationRepository = notificationRepository;
        this.userRepository = userRepository;
        this.projectRepository = projectRepository;
        this.schedulerService = schedulerService;
        this.notificationConverter = notificationConverter;
        this.notificationStateEventPublisher = eventPublisher;
    }

    @Transactional(readOnly = true)
    public FcmNotifications getAllNotifications() {
        List<Notification> notifications = notificationRepository.findAll();
        return new FcmNotifications()
                .setNotifications(notificationConverter.entitiesToDtos(notifications));
    }

    @Transactional(readOnly = true)
    public FcmNotificationDto getNotificationById(long id) {
        Optional<Notification> notification = notificationRepository.findById(id);
        return notificationConverter.entityToDto(notification.orElseGet(Notification::new));
    }

    @Transactional(readOnly = true)
    public FcmNotifications getNotificationsBySubjectId(String subjectId) {
        Optional<User> user = this.userRepository.findBySubjectId(subjectId);
        if (user.isEmpty()) {
            throw new NotFoundException(INVALID_SUBJECT_ID_MESSAGE);
        }
        List<Notification> notifications = notificationRepository.findByUserId(user.get().getId());
        return new FcmNotifications()
                .setNotifications(notificationConverter.entitiesToDtos(notifications));
    }

    @Transactional(readOnly = true)
    public FcmNotifications getNotificationsByProjectIdAndSubjectId(
            String projectId, String subjectId) {
        User user = subjectAndProjectExistElseThrow(subjectId, projectId);

        List<Notification> notifications = notificationRepository.findByUserId(user.getId());
        return new FcmNotifications()
                .setNotifications(notificationConverter.entitiesToDtos(notifications));
    }

    @Transactional(readOnly = true)
    public FcmNotifications getNotificationsByProjectId(String projectId) {
        Optional<Project> project = projectRepository.findByProjectId(projectId);

        if (project.isEmpty()) {
            throw new NotFoundException("Project not found with projectId " + projectId);
        }
        List<User> users = this.userRepository.findByProjectId(project.get().getId());
        Set<Notification> notifications = new HashSet<>();
        users.stream()
                .map((User user) -> this.notificationRepository.findByUserId(user.getId()))
                .forEach(notifications::addAll);
        return new FcmNotifications()
                .setNotifications(notificationConverter.entitiesToDtos(notifications));
    }

    @Transactional(readOnly = true)
    public boolean checkIfNotificationExists(FcmNotificationDto notificationDto, String subjectId) {
        Optional<User> user = this.userRepository.findBySubjectId(subjectId);
        if (user.isEmpty()) {
            throw new NotFoundException(INVALID_SUBJECT_ID_MESSAGE);
        }
        Notification notification = new Notification.NotificationBuilder(notificationConverter.dtoToEntity(notificationDto)).user(user.get()).build();

        List<Notification> notifications = this.notificationRepository.findByUserId(user.get().getId());
        return notifications.contains(notification);
    }

    // TODO : WIP
    @Transactional(readOnly = true)
    public FcmNotifications getFilteredNotifications(
            String type,
            boolean delivered,
            int ttlSeconds,
            LocalDateTime startTime,
            LocalDateTime endTime,
            int limit) {
        return null;
    }

    @Transactional
    public FcmNotificationDto addNotification(
            FcmNotificationDto notificationDto, String subjectId, String projectId, boolean schedule) {

        User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        Optional<Notification> notification = notificationRepository
                .findByUserIdAndSourceIdAndScheduledTimeAndTitleAndBodyAndTypeAndTtlSeconds(
                        user.getId(),
                        notificationDto.getSourceId(),
                        notificationDto.getScheduledTime(),
                        notificationDto.getTitle(),
                        notificationDto.getBody(),
                        notificationDto.getType(),
                        notificationDto.getTtlSeconds());

        if (notification.isEmpty()) {
            Notification notificationSaved =
                    this.notificationRepository.saveAndFlush(
                            new Notification.NotificationBuilder(notificationConverter.dtoToEntity(notificationDto)).user(user).build());
            user.getUsermetrics().setLastOpened(Instant.now());
            this.userRepository.save(user);
            addNotificationStateEvent(
                    notificationSaved, MessageState.ADDED, notificationSaved.getCreatedAt().toInstant());
            if (schedule) {
                this.schedulerService.schedule(notificationSaved);
            }
            return notificationConverter.entityToDto(notificationSaved);
        } else {
            throw new NotificationAlreadyExistsException(
                    "The Notification Already exists. Please Use update endpoint",
                    notificationConverter.entityToDto(notification.get()));
        }
    }

    @Transactional
    public FcmNotificationDto addNotification(
            FcmNotificationDto notificationDto, String subjectId, String projectId) {

        User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        Optional<Notification> notification = notificationRepository
                .findByUserIdAndSourceIdAndScheduledTimeAndTitleAndBodyAndTypeAndTtlSeconds(
                        user.getId(),
                        notificationDto.getSourceId(),
                        notificationDto.getScheduledTime(),
                        notificationDto.getTitle(),
                        notificationDto.getBody(),
                        notificationDto.getType(),
                        notificationDto.getTtlSeconds());

        if (notification.isEmpty()) {
            Notification notificationSaved =
                    this.notificationRepository.saveAndFlush(
                            new Notification.NotificationBuilder(notificationConverter.dtoToEntity(notificationDto)).user(user).build());
            user.getUsermetrics().setLastOpened(Instant.now());
            this.userRepository.save(user);
            addNotificationStateEvent(
                    notificationSaved, MessageState.ADDED, notificationSaved.getCreatedAt().toInstant());
            this.schedulerService.schedule(notificationSaved);
            return notificationConverter.entityToDto(notificationSaved);
        } else {
            throw new NotificationAlreadyExistsException(
                    "The Notification Already exists. Please Use update endpoint",
                    notificationConverter.entityToDto(notification.get()));
        }
    }

    private void addNotificationStateEvent(
            Notification notification, MessageState state, Instant time) {
        if (notificationStateEventPublisher != null) {
            NotificationStateEventDto notificationStateEvent =
                    new NotificationStateEventDto(this, notification, state, null, time);
            notificationStateEventPublisher.publishEvent(notificationStateEvent);
        }
    }


    @Transactional
    public FcmNotificationDto updateNotification(
            FcmNotificationDto notificationDto, String subjectId, String projectId) {

        if (notificationDto.getId() == null) {
            throw new InvalidNotificationDetailsException(
                    "ID must be supplied for updating the notification");
        }

        User user = subjectAndProjectExistElseThrow(subjectId, projectId);

        Optional<Notification> notification =
                this.notificationRepository.findById(notificationDto.getId());

        if (notification.isEmpty()) {
            throw new NotFoundException("Notification does not exist. Please create first");
        }

        Notification newNotification = new Notification.NotificationBuilder(notification.get())
                .body(notificationDto.getBody())
                .scheduledTime(notificationDto.getScheduledTime())
                .sourceId(notificationDto.getSourceId())
                .title(notificationDto.getTitle())
                .ttlSeconds(notificationDto.getTtlSeconds())
                .type(notificationDto.getType())
                .user(user)
                .fcmMessageId(String.valueOf(notificationDto.hashCode()))
                .build();
        Notification notificationSaved = this.notificationRepository.saveAndFlush(newNotification);
        addNotificationStateEvent(
                notificationSaved, MessageState.UPDATED, notificationSaved.getUpdatedAt().toInstant());
        if (!notification.get().isDelivered()) {
            this.schedulerService.updateScheduled(notificationSaved);
        }
        return notificationConverter.entityToDto(notificationSaved);
    }

    @Transactional
    public FcmNotifications scheduleAllUserNotifications(String subjectId, String projectId) {

        User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        List<Notification> notifications = notificationRepository.findByUserId(user.getId());
        this.schedulerService.scheduleMultiple(notifications);
        return new FcmNotifications()
                .setNotifications(notificationConverter.entitiesToDtos(notifications));
    }

    @Transactional
    public FcmNotificationDto scheduleNotification(String subjectId, String projectId, long notificationId) {

        User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        Optional<Notification> notification = notificationRepository.findByIdAndUserId(notificationId, user.getId());
        if (notification.isEmpty()) {
            throw new NotFoundException(
                    "The Notification with Id "
                            + notificationId
                            + " does not exist in project "
                            + projectId
                            + " for user "
                            + subjectId);
        }
        this.schedulerService.schedule(notification.get());
        return notificationConverter.entityToDto(notification.get());
    }

    @Transactional
    public void removeNotificationsForUser(String projectId, String subjectId) {
        User user = subjectAndProjectExistElseThrow(subjectId, projectId);

        List<Notification> notifications = this.notificationRepository.findByUserId(user.getId());
        this.schedulerService.deleteScheduledMultiple(notifications);

        this.notificationRepository.deleteByUserId(user.getId());
    }

    @Transactional
    public void updateDeliveryStatus(String fcmMessageId, boolean isDelivered) {
        Optional<Notification> notification =
                this.notificationRepository.findByFcmMessageId(fcmMessageId);

        notification.ifPresentOrElse(
                (Notification n) -> {

                    Notification newNotif = new Notification.NotificationBuilder(n).delivered(isDelivered).build();
                    this.notificationRepository.save(newNotif);
                },
                () -> {
                    throw new InvalidNotificationDetailsException(
                            "Notification with the provided FCM message ID does not exist.");
                });
    }

    // TODO: Investigate if notifications can be marked in the state CANCELLED when deleted.
    @Transactional
    public void deleteNotificationByProjectIdAndSubjectIdAndNotificationId(String projectId, String subjectId, Long id) {
        User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        Long userId = user.getId();

        if (this.notificationRepository.existsByIdAndUserId(id, userId)) {
            this.schedulerService.deleteScheduled(
                    this.notificationRepository.findByIdAndUserId(id, userId).get());
            this.notificationRepository.deleteByIdAndUserId(id, userId);
        } else
            throw new InvalidNotificationDetailsException(
                    "Notification with the provided ID does not exist.");
    }

    @Transactional
    public void removeNotificationsForUserUsingFcmToken(String fcmToken) {
        Optional<User> user = this.userRepository.findByFcmToken(fcmToken);

        user.ifPresentOrElse(
                (User user1) -> {
                    this.schedulerService.deleteScheduledMultiple(
                            this.notificationRepository.findByUserId(user1.getId()));

                    this.notificationRepository.deleteByUserId(user1.getId());
                },
                () -> {
                    throw new InvalidUserDetailsException("The user with the given Fcm Token does not exist");
                });
    }

    @Transactional
    public FcmNotifications addNotifications(
            FcmNotifications notificationDtos, String subjectId, String projectId, boolean schedule) {
        final User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        List<Notification> notifications = notificationRepository.findByUserId(user.getId());

        List<Notification> newNotifications =
                notificationDtos.getNotifications().stream()
                        .map(notificationConverter::dtoToEntity)
                        .map(n -> new Notification.NotificationBuilder(n).user(user).build())
                        .filter(notification -> !notifications.contains(notification))
                        .collect(Collectors.toList());

        List<Notification> savedNotifications = this.notificationRepository.saveAll(newNotifications);
        this.notificationRepository.flush();
        savedNotifications.forEach(
                n -> addNotificationStateEvent(n, MessageState.ADDED, n.getCreatedAt().toInstant()));
        if (schedule) {
            this.schedulerService.scheduleMultiple(savedNotifications);
        }
        return new FcmNotifications()
                .setNotifications(notificationConverter.entitiesToDtos(savedNotifications));
    }

    @Transactional
    public List<Notification> addNotifications(List<Notification> notifications,User user) {
        List<Notification> newNotifications =
                notifications.stream()
                        .filter(notification ->
                            notificationRepository
                                    .findByUserIdAndSourceIdAndScheduledTimeAndTitleAndBodyAndTypeAndTtlSeconds(
                                            user.getId(),
                                            notification.getSourceId(),
                                            notification.getScheduledTime(),
                                            notification.getTitle(),
                                            notification.getBody(),
                                            notification.getType(),
                                            notification.getTtlSeconds()).isPresent()
                        )
                        .collect(Collectors.toList());

        List<Notification> savedNotifications = this.notificationRepository.saveAll(newNotifications);
        this.notificationRepository.flush();
        savedNotifications.forEach(
                n -> addNotificationStateEvent(n, MessageState.ADDED, n.getCreatedAt().toInstant()));
        this.schedulerService.scheduleMultiple(savedNotifications);
        return savedNotifications;
    }

    @Transactional
    public FcmNotifications addNotifications(
            FcmNotifications notificationDtos, String subjectId, String projectId) {
        final User user = subjectAndProjectExistElseThrow(subjectId, projectId);
        List<Notification> notifications = notificationRepository.findByUserId(user.getId());

        List<Notification> newNotifications =
                notificationDtos.getNotifications().stream()
                        .map(notificationConverter::dtoToEntity)
                        .map(n -> new Notification.NotificationBuilder(n).user(user).build())
                        .filter(notification ->
                                notificationRepository
                                        .findByUserIdAndSourceIdAndScheduledTimeAndTitleAndBodyAndTypeAndTtlSeconds(
                                                user.getId(),
                                                notification.getSourceId(),
                                                notification.getScheduledTime(),
                                                notification.getTitle(),
                                                notification.getBody(),
                                                notification.getType(),
                                                notification.getTtlSeconds()).isPresent()
                        )
                        .collect(Collectors.toList());

        List<Notification> savedNotifications = this.notificationRepository.saveAll(newNotifications);
        this.notificationRepository.flush();
        savedNotifications.forEach(
                n -> addNotificationStateEvent(n, MessageState.ADDED, n.getCreatedAt().toInstant()));
        this.schedulerService.scheduleMultiple(savedNotifications);
        return new FcmNotifications()
                .setNotifications(notificationConverter.entitiesToDtos(savedNotifications));
    }

    public User subjectAndProjectExistElseThrow(String subjectId, String projectId) {
        Optional<Project> project = this.projectRepository.findByProjectId(projectId);
        if (project.isEmpty()) {
            throw new NotFoundException(
                    "Project Id does not exist. Please create a project with the ID first");
        }

        Optional<User> user =
                this.userRepository.findBySubjectIdAndProjectId(subjectId, project.get().getId());
        if (user.isEmpty()) {
            throw new NotFoundException(INVALID_SUBJECT_ID_MESSAGE);
        }

        return user.get();
    }

    @Transactional(readOnly = true)
    public Notification getNotificationByProjectIdAndSubjectIdAndNotificationId(
            String projectId, String subjectId, long notificationId) {
        User user = subjectAndProjectExistElseThrow(subjectId, projectId);

        Optional<Notification> notification =
                notificationRepository.findByIdAndUserId(notificationId, user.getId());

        if (notification.isEmpty()) {
            throw new InvalidNotificationDetailsException(
                    "The Notification with Id "
                            + notificationId
                            + " does not exist in project "
                            + projectId
                            + " for user "
                            + subjectId);
        }
        return notification.get();
    }

    @Transactional(readOnly = true)
    public Notification getNotificationByMessageId(String messageId) {
        Optional<Notification> notification = this.notificationRepository.findByFcmMessageId(messageId);
        if (notification.isEmpty()) {
            throw new InvalidNotificationDetailsException(
                    "The Notification with FCM Message Id " + messageId + "does not exist.");
        }
        return notification.get();
    }

}

File: src/main/java/org/radarbase/appserver/service/FcmNotificationService.kt
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */
package org.radarbase.appserver.service

import org.radarbase.appserver.dto.fcm.FcmNotificationDto
import org.radarbase.appserver.dto.fcm.FcmNotifications
import org.radarbase.appserver.entity.Notification
import org.radarbase.appserver.entity.Notification.NotificationBuilder
import org.radarbase.appserver.entity.Task
import org.radarbase.appserver.entity.User
import org.radarbase.appserver.event.state.MessageState
import org.radarbase.appserver.event.state.dto.NotificationStateEventDto
import org.radarbase.appserver.exception.InvalidNotificationDetailsException
import org.radarbase.appserver.exception.InvalidUserDetailsException
import org.radarbase.appserver.exception.NotFoundException
import org.radarbase.appserver.exception.NotificationAlreadyExistsException
import org.radarbase.appserver.mapper.NotificationMapper
import org.radarbase.appserver.repository.NotificationRepository
import org.radarbase.appserver.repository.ProjectRepository
import org.radarbase.appserver.repository.UserRepository
import org.radarbase.appserver.service.scheduler.MessageSchedulerService
import org.radarbase.appserver.util.checkInvalidDetails
import org.radarbase.appserver.util.checkPresence
import org.springframework.context.ApplicationEventPublisher
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.time.LocalDateTime

/**
 * [Service] for interacting with the [Notification] [jakarta.persistence.Entity]
 * using the [NotificationRepository].
 *
 * @author yatharthranjan
 */
@Suppress("unused")
@Service
class FcmNotificationService(
    private val notificationRepository: NotificationRepository,
    private val userRepository: UserRepository,
    private val projectRepository: ProjectRepository,
    private val schedulerService: MessageSchedulerService<Notification>,
    private val notificationConverter: NotificationMapper,
    private val notificationStateEventPublisher: ApplicationEventPublisher?
) : NotificationService {

    // TODO Add option to specify a scheduling provider (default will be fcm)
    // TODO: Use strategy pattern for handling notifications for scheduling and adding to database

    @Transactional(readOnly = true)
    fun getAllNotifications(): FcmNotifications {
        val notifications: List<Notification> = notificationRepository.findAll()
        return FcmNotifications()
            .withNotifications(notificationConverter.entitiesToDtos(notifications))
    }

    @Transactional(readOnly = true)
    fun getNotificationById(id: Long): FcmNotificationDto {
        val notification: Notification? = notificationRepository.findByIdOrNull(id)
        return notificationConverter.entityToDto(notification ?: Notification())
    }

    @Transactional(readOnly = true)
    fun getNotificationsBySubjectId(subjectId: String?): FcmNotifications {
        val user = this.userRepository.findBySubjectId(subjectId)
        if (user == null) {
            throw NotFoundException(INVALID_SUBJECT_ID_MESSAGE)
        }
        val notifications: List<Notification> = notificationRepository.findByUserId(user.id)
        return FcmNotifications()
            .withNotifications(notificationConverter.entitiesToDtos(notifications))
    }

    @Transactional(readOnly = true)
    fun getNotificationsByProjectIdAndSubjectId(
        projectId: String?, subjectId: String?
    ): FcmNotifications {
        return subjectAndProjectExistElseThrow(subjectId, projectId).let { user ->
            notificationRepository.findByUserId(user.id)
        }.let { notifications ->
            FcmNotifications()
                .withNotifications(notificationConverter.entitiesToDtos(notifications))
        }
    }

    @Transactional(readOnly = true)
    fun getNotificationsByProjectId(projectId: String?): FcmNotifications {
        return checkPresence(projectRepository.findByProjectId(projectId)) {
            "Project not found with projectId $projectId"
        }.let { project ->
            this.userRepository.findByProjectId(project.id)
        }.let { users ->
            hashSetOf<Notification>().also { notifications ->
                users.map { user ->
                    notificationRepository.findByUserId(user.id)
                }.forEach { userNotifications: List<Notification> ->
                    notifications.addAll(userNotifications)
                }
            }
        }.let { notifications ->
            FcmNotifications()
                .withNotifications(notificationConverter.entitiesToDtos(notifications))
        }
    }

    @Transactional(readOnly = true)
    fun checkIfNotificationExists(notificationDto: FcmNotificationDto, subjectId: String?): Boolean {
        checkPresence(this.userRepository.findBySubjectId(subjectId)) {
            INVALID_SUBJECT_ID_MESSAGE
        }.let { user ->
            val notification =
                NotificationBuilder(notificationConverter.dtoToEntity(notificationDto)).user(user).build()
            val notifications: List<Notification> = this.notificationRepository.findByUserId(user.id)
            return notifications.contains(notification)
        }
    }

    // TODO : WIP
    @Transactional(readOnly = true)
    fun getFilteredNotifications(
        type: String?,
        delivered: Boolean,
        ttlSeconds: Int,
        startTime: LocalDateTime?,
        endTime: LocalDateTime?,
        limit: Int
    ): FcmNotifications? = null

    @Transactional
    fun addNotification(
        notificationDto: FcmNotificationDto, subjectId: String?, projectId: String?, schedule: Boolean
    ): FcmNotificationDto {
        val user = subjectAndProjectExistElseThrow(subjectId, projectId)
        val notification: Notification? = notificationRepository
            .findByUserIdAndSourceIdAndScheduledTimeAndTitleAndBodyAndTypeAndTtlSeconds(
                user.id,
                notificationDto.sourceId,
                notificationDto.scheduledTime,
                notificationDto.title,
                notificationDto.body,
                notificationDto.type,
                notificationDto.ttlSeconds
            )

        if (notification == null) {
            val notificationSaved = this.notificationRepository.saveAndFlush<Notification>(
                NotificationBuilder(notificationConverter.dtoToEntity(notificationDto)).user(user).build()
            )
            user.usermetrics!!.lastOpened = Instant.now()
            this.userRepository.save(user)
            addNotificationStateEvent(
                notificationSaved, MessageState.ADDED, notificationSaved.createdAt!!.toInstant()
            )
            if (schedule) {
                this.schedulerService.schedule(notificationSaved)
            }
            return notificationConverter.entityToDto(notificationSaved)
        } else {
            throw NotificationAlreadyExistsException(
                "The Notification Already exists. Please Use update endpoint",
                notificationConverter.entityToDto(notification)
            )
        }
    }

    @Transactional
    fun addNotification(
        notificationDto: FcmNotificationDto, subjectId: String?, projectId: String?
    ): FcmNotificationDto {
        val user = subjectAndProjectExistElseThrow(subjectId, projectId)
        val notification: Notification? = notificationRepository
            .findByUserIdAndSourceIdAndScheduledTimeAndTitleAndBodyAndTypeAndTtlSeconds(
                user.id,
                notificationDto.sourceId,
                notificationDto.scheduledTime,
                notificationDto.title,
                notificationDto.body,
                notificationDto.type,
                notificationDto.ttlSeconds
            )

        if (notification == null) {
            val notificationSaved = this.notificationRepository.saveAndFlush<Notification>(
                NotificationBuilder(notificationConverter.dtoToEntity(notificationDto)).user(user).build()
            )
            user.usermetrics!!.lastOpened = Instant.now()
            this.userRepository.save(user)
            addNotificationStateEvent(
                notificationSaved, MessageState.ADDED, notificationSaved.createdAt!!.toInstant()
            )
            this.schedulerService.schedule(notificationSaved)
            return notificationConverter.entityToDto(notificationSaved)
        } else {
            throw NotificationAlreadyExistsException(
                "The Notification Already exists. Please Use update endpoint",
                notificationConverter.entityToDto(notification)
            )
        }
    }

    private fun addNotificationStateEvent(
        notification: Notification, state: MessageState, time: Instant
    ) {
        if (notificationStateEventPublisher != null) {
            val notificationStateEvent =
                NotificationStateEventDto(this, notification, state, null, time)
            notificationStateEventPublisher.publishEvent(notificationStateEvent)
        }
    }


    @Transactional
    fun updateNotification(
        notificationDto: FcmNotificationDto, subjectId: String?, projectId: String?
    ): FcmNotificationDto {
        val notificationId = notificationDto.id

        checkInvalidDetails<InvalidNotificationDetailsException>({ notificationId == null }, {
            "ID must be supplied for updating the notification"
        })

        val user = subjectAndProjectExistElseThrow(subjectId, projectId)

        val notification = checkPresence(this.notificationRepository.findByIdOrNull(notificationId)) {
            throw NotFoundException("Notification does not exist. Please create first")
        }

        val newNotification = NotificationBuilder(notification)
            .body(notificationDto.body)
            .scheduledTime(notificationDto.scheduledTime)
            .sourceId(notificationDto.sourceId)
            .title(notificationDto.title)
            .ttlSeconds(notificationDto.ttlSeconds)
            .type(notificationDto.type)
            .user(user)
            .fcmMessageId(notificationDto.hashCode().toString())
            .build()
        val notificationSaved = this.notificationRepository.saveAndFlush<Notification>(newNotification)

        addNotificationStateEvent(
            notificationSaved, MessageState.UPDATED, notificationSaved.updatedAt!!.toInstant()
        )
        if (!notification.delivered) {
            this.schedulerService.updateScheduled(notificationSaved)
        }
        return notificationConverter.entityToDto(notificationSaved)
    }

    @Transactional
    fun scheduleAllUserNotifications(subjectId: String, projectId: String): FcmNotifications {
        val user = subjectAndProjectExistElseThrow(subjectId, projectId)
        val notifications: List<Notification> = notificationRepository.findByUserId(user.id)
        this.schedulerService.scheduleMultiple(notifications)
        return FcmNotifications()
            .withNotifications(notificationConverter.entitiesToDtos(notifications))
    }

    @Transactional
    fun scheduleNotification(subjectId: String?, projectId: String?, notificationId: Long): FcmNotificationDto {
        val user = subjectAndProjectExistElseThrow(subjectId, projectId)
        val notification = notificationRepository.findByIdAndUserId(notificationId, user.id!!)
        checkPresence(notification) {
            "The Notification with Id $notificationId does not exist in project $projectId for user $subjectId"
        }
        this.schedulerService.schedule(notification)
        return notificationConverter.entityToDto(notification)
    }

    @Transactional
    fun removeNotificationsForUser(projectId: String?, subjectId: String?) {
        val user = subjectAndProjectExistElseThrow(subjectId, projectId)

        val notifications: List<Notification> = this.notificationRepository.findByUserId(user.id)
        this.schedulerService.deleteScheduledMultiple(notifications)

        this.notificationRepository.deleteByUserId(user.id)
    }

    @Transactional
    fun updateDeliveryStatus(fcmMessageId: String?, isDelivered: Boolean) {
        val notification =
            this.notificationRepository.findByFcmMessageId(fcmMessageId)

        checkInvalidDetails<InvalidNotificationDetailsException>(
            { notification == null },
            {
                "Notification with the provided FCM message ID does not exist."
            }
        )
        val newNotif = NotificationBuilder(notification).delivered(isDelivered).build()
        this.notificationRepository.save<Notification?>(newNotif)
    }

    // TODO: Investigate if notifications can be marked in the state CANCELLED when deleted.
    @Transactional
    fun deleteNotificationByProjectIdAndSubjectIdAndNotificationId(projectId: String?, subjectId: String?, id: Long) {
        val user = subjectAndProjectExistElseThrow(subjectId, projectId)
        val userId = user.id

        if (this.notificationRepository.existsByIdAndUserId(id, userId)) {
            this.schedulerService.deleteScheduled(
                this.notificationRepository.findByIdAndUserId(id, userId!!)
            )
            this.notificationRepository.deleteByIdAndUserId(id, userId)
        } else throw InvalidNotificationDetailsException(
            "Notification with the provided ID does not exist."
        )
    }

    @Transactional
    fun removeNotificationsForUserUsingTaskId(projectId: String?, subjectId: String?, taskId: Long?) {
        val user = subjectAndProjectExistElseThrow(subjectId, projectId)

        val notifications: List<Notification> = this.notificationRepository.findByUserIdAndTaskId(user.id, taskId)
        this.schedulerService.deleteScheduledMultiple(notifications)

        this.notificationRepository.deleteByUserIdAndTaskId(user.id, taskId)
    }

    @Transactional
    fun removeNotificationsForUserUsingFcmToken(fcmToken: String?) {
        val user = this.userRepository.findByFcmToken(fcmToken)

        checkInvalidDetails<InvalidUserDetailsException>(
            { user == null },
            { "The user with the given Fcm Token does not exist" }
        )


        this.schedulerService.deleteScheduledMultiple(
            this.notificationRepository.findByUserId(user!!.id)
        )

        this.notificationRepository.deleteByUserId(user.id)
    }

    @Transactional
    fun deleteNotificationsByTaskId(task: Task?) {
        val taskId = task?.id ?: return
        if (notificationRepository.existsByTaskId(taskId)) {
            val notifications: List<Notification> = notificationRepository.findByTaskId(taskId)
            schedulerService.deleteScheduledMultiple(notifications)
            notificationRepository.deleteByTaskId(taskId)
        }
    }

    @Transactional
    fun addNotifications(
        notificationDtos: FcmNotifications, subjectId: String?, projectId: String?, schedule: Boolean
    ): FcmNotifications {

        val newNotifications: List<Notification> = subjectAndProjectExistElseThrow(subjectId, projectId).let { user ->
            notificationRepository.findByUserId(user.id).let { notifications ->
                notificationDtos.notifications.map { dto: FcmNotificationDto ->
                    notificationConverter.dtoToEntity(dto)
                }.map { notification ->
                    NotificationBuilder(notification).user(user).build()
                }.filter { notification ->
                    !notifications.contains(notification)
                }
            }
        }

        val savedNotifications = this.notificationRepository.saveAll(newNotifications)
        this.notificationRepository.flush()
        savedNotifications.forEach { n: Notification ->
            addNotificationStateEvent(
                n,
                MessageState.ADDED,
                n.createdAt!!.toInstant()
            )
        }

        if (schedule) {
            this.schedulerService.scheduleMultiple(savedNotifications)
        }
        return FcmNotifications()
            .withNotifications(notificationConverter.entitiesToDtos(savedNotifications))
    }

    @Transactional
    fun addNotifications(notifications: List<Notification>?, user: User): List<Notification> {
        notifications ?: return listOf()
        val newNotifications: List<Notification> = notifications.filter { notification: Notification ->
            notificationRepository
                .findByUserIdAndSourceIdAndScheduledTimeAndTitleAndBodyAndTypeAndTtlSeconds(
                    user.id,
                    notification.sourceId,
                    notification.scheduledTime,
                    notification.title,
                    notification.body,
                    notification.type,
                    notification.ttlSeconds
                ) == null
        }

        val savedNotifications: List<Notification> = this.notificationRepository.saveAllAndFlush(newNotifications)
        savedNotifications.forEach { n: Notification? ->
            addNotificationStateEvent(
                n!!,
                MessageState.ADDED,
                n.createdAt!!.toInstant()
            )
        }
        this.schedulerService.scheduleMultiple(savedNotifications)
        return savedNotifications
    }

    @Transactional
    fun addNotifications(
        notificationDtos: FcmNotifications, subjectId: String?, projectId: String?
    ): FcmNotifications {

        val newNotifications: List<Notification> = subjectAndProjectExistElseThrow(subjectId, projectId).let { user ->
            notificationRepository.findByUserId(user.id).let { notifications ->
                notificationDtos.notifications.map { dto: FcmNotificationDto ->
                    notificationConverter.dtoToEntity(dto)
                }.map { notification ->
                    NotificationBuilder(notification).user(user).build()
                }.filter { notification ->
                    !notifications.contains(notification)
                }
            }
        }

        val savedNotifications = this.notificationRepository.saveAll(newNotifications)
        this.notificationRepository.flush()
        savedNotifications.forEach { n: Notification ->
            addNotificationStateEvent(
                n,
                MessageState.ADDED,
                n.createdAt!!.toInstant()
            )
        }

        this.schedulerService.scheduleMultiple(savedNotifications)
        return FcmNotifications()
            .withNotifications(notificationConverter.entitiesToDtos(savedNotifications))
    }

    fun subjectAndProjectExistElseThrow(subjectId: String?, projectId: String?): User {
        return checkPresence(this.projectRepository.findByProjectId(projectId)) {
            "Project Id does not exist. Please create a project with the ID first"
        }.let { project ->
            checkPresence(this.userRepository.findBySubjectIdAndProjectId(subjectId, project.id)) {
                INVALID_SUBJECT_ID_MESSAGE
            }
        }
    }

    @Transactional(readOnly = true)
    fun getNotificationByProjectIdAndSubjectIdAndNotificationId(
        projectId: String?, subjectId: String?, notificationId: Long
    ): Notification {
        val user = subjectAndProjectExistElseThrow(subjectId, projectId)

        val notification = notificationRepository.findByIdAndUserId(notificationId, user.id!!)

        checkInvalidDetails<InvalidNotificationDetailsException>({
            notification == null
        }, {
            "The Notification with Id $notificationId does not exist in project $projectId for user $subjectId"
        })

        return notification!!
    }

    @Transactional(readOnly = true)
    fun getNotificationByMessageId(messageId: String?): Notification {
        val notification = this.notificationRepository.findByFcmMessageId(messageId)
        checkInvalidDetails<InvalidNotificationDetailsException>({
            notification == null
        }, {
            "The Notification with FCM Message Id $messageId does not exist."
        })
        return notification!!
    }

    companion object {
        private const val INVALID_SUBJECT_ID_MESSAGE = "The supplied Subject ID is invalid. No user found. Please Create a User First."
    }
}
